store/tikv/lock_resolver.go

```diff
@@ -578,10 +595,136 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
+		curKeys := keys
+
+		go func() {
+			for {
+				checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+					Keys:         curKeys,
+					StartVersion: l.TxnID,
+				}
+				req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+				tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+				resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				regionErr, err := resp.GetRegionError()
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				if regionErr != nil {
+					err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+					if err != nil {
+						errChan <- errors.Trace(err)
+						return
+					}
+					continue
+				}
+				if resp.Resp == nil {
+					errChan <- errors.Trace(ErrBodyMissing)
+					return
+				}
+				checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+
+				// Check each lock.
+				if len(checkResp.Locks) < len(curKeys) {
+					// A lock is missing - the transaction must either be rolled back or committed.
+					commitTsMutex.Lock()
+					// I recommend to add a check if the resp.CommitTs equals to commitTs when missingLock is already true. This may help us to find bugs
+					if !missingLock {
+						commitTs = checkResp.CommitTs
+					}
+					missingLock = true
+					commitTsMutex.Unlock()
+
+					if commitTs != checkResp.CommitTs {
+						errChan <- errors.Errorf("commit TS mismatch in async commit recovery: %v and %v", commitTs, checkResp.CommitTs)
+					} else {
+						errChan <- nil
+					}
+					return
+				}
+
+				for _, lockInfo := range checkResp.Locks {
+
+					if lockInfo.LockVersion != l.TxnID {
+						err = errors.Errorf("unexpected timestamp, expected: %v, found: %v", l.TxnID, lockInfo.LockVersion)
+						logutil.BgLogger().Error("resolveLockAsync error", zap.Error(err))
+						errChan <- err
+						return
+					}
+
+					commitTsMutex.Lock()
+					if !missingLock && lockInfo.MinCommitTs > commitTs {
+						commitTs = lockInfo.MinCommitTs
+					}
+					locks = append(locks, NewLock(lockInfo))
+					commitTsMutex.Unlock()
+				}
+				break
+			}
+			errChan <- nil
+		}()
+	}
+
+	var errs []string
+	for range regions {
+		err1 := <-errChan
+		if err1 != nil {
+			errs = append(errs, err1.Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		return errors.Errorf("async commit recovery (sending CheckSecondaryLocks) finished with errors: %v", errs)
+	}
+
+	// TODO(nrc, cfzjywxk) schema lease check
+
+	status.commitTS = commitTs
+
+	// Resolve every lock in the transaction.
+	// TODO(nrc) it should be possible to resolve locks in parallel, however, it is unsafe to call resolveLock concurrently
+	// because cleanRegions might be mutated by multiple threads.
+	for _, lock := range locks {
+		if lock == nil {
+			errChan <- nil
+			continue
+		}
+		err = lr.resolveLock(bo, lock, status, lite, cleanRegions)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
```

Comment:
This should work but can be improved.
Because we know all the keys and their regions, locks in the same region can be resolved by a single request, and the `lite` option can be always true for async commit.

Reply:
Why is it always ok to use the lite option?

Reply:
Non-lite means that we scan the whole region for locks of a certain transaction. But for async commit transactions, we know the list of all locks, so we can directly specify the keys to resolve.